src/Terminal/commands/grep.ts

```diff
@@ -0,0 +1,179 @@
+    if (!script) {
+      notFiles.push(arg);
+    }
+    return script;
+  });
+  return [files, notFiles];
+}
+
+function filterOpts([options, otherArgs]: [Options, string[]], arg: string): [Options, string[]] {
+  const isOption: boolean = Object.keys(OK_ARGS).some((key: string): boolean =>
+    OK_ARGS[key as keyof OkArgs].includes(arg) ? (options[key as keyof Options] = true) : false,
+  );
+  return isOption ? [options, otherArgs] : [options, [...otherArgs, arg]];
+}
+
+function filterArgs(args: string[]): [Options, string[], FilePath | null] {
+  const initOpts: Options = {
+    lineNum: false,
+    regExpr: false,
+    yesName: false,
+    notName: false,
+    multiFile: false,
+    toFile: false,
+    overWrt: false,
+    quiet: false,
+  };
+
+  const outputArgIndex = OK_ARGS.toFile.reduce((ret: number, arg: string) => {
+    const argIndex = args.indexOf(arg);
+    return argIndex > -1 ? argIndex : ret;
+  }, NaN);
+  // if outputArgIndex !NaN grab next arg as output file
+  const outFileStr: string | undefined = !isNaN(outputArgIndex) ? args.splice(outputArgIndex + 1)?.[0] : undefined;
+
+  const outFile = outFileStr ? Terminal.getFilepath(outFileStr) : null;
+
+  const [options, otherArgs]: [Options, string[]] = args.map(String).reduce(filterOpts, [initOpts, []]);
+
+  options.toFile = !!outFile;
+
+  return [options, otherArgs, outFile];
+}
+
+export function grep(args: (string | number | boolean)[], server: BaseServer): void {
+  if (!args.length) {
+    return Terminal.error("Incorrect usage of grep command. Usage: grep [OPTION]... PATTERN [FILE]...");
+  }
+
+  const [options, otherArgs, outFilePath]: [Options, string[], FilePath | null] = filterArgs(args.map(String));
+  if (!outFilePath || !hasTextExtension(outFilePath)) {
+    return Terminal.error(`grep failed: Invalid output file "${outFilePath}". Output file must be a text file.`);
+  }
+
+  console.log(options);
+  if (options.toFile && !options.overWrt) {
+    const overWrtFileExists = [...allContentFiles(server)]
+      .map((tuple: FileTuple): string => tuple[1].filename)
+      .includes(outFilePath);
+    if (overWrtFileExists) {
+      return Terminal.error(
+        `grep failed: Invalid output file "${outFilePath}". Output file must not already exist. Pass -f to force.`,
+      );
+    }
+  }
+
+  const fileArgs = otherArgs.slice(1);
+
+  const [files, notFiles]: [(ContentFile | null)[], string[]] = fileArgs.length
+    ? getArgFiles(fileArgs)
+    : getServerFiles(server);
+
+  options.multiFile = files.length > 1;
+
+  if (notFiles.length) {
+    return Terminal.error(`Invalid filename(s): ${notFiles.join(", ")}`);
+  }
+
+  try {
+    const pattern: string | RegExp = options.regExpr ? new RegExp(otherArgs[0], "g") : otherArgs[0];
+    const parseFunc: PtLineParser = getParseFunc.bind(null, pattern, options);
+    const result: string = files
+      .flatMap(parseFile(parseFunc))
+      .filter((line: string) => line.length)
```

Comment:
If you're going to filter unmatched lines here, why even bother including them in the result of `parseFile`?
For large files you're creating a large array only to discard most of it.